src/pages/portfolio/ui/assets-table.tsx

```diff
@@ -0,0 +1,450 @@
+    // Offset the hue by the hash value to make UM naturally orange
+    return (num * 83) % 360;
+  };
+
+  const distributionWithMetadata = await Promise.all(
+    valuesWithMetadata.map(async ({ value, balance, hasError }) => {
+      const metadata = getMetadataFromBalancesResponse.optional(balance);
+      const assetIdHex = metadata?.penumbraAssetId?.inner.toString() ?? '';
+
+      // Use Web Crypto API to hash the hex string
+      const hashBuffer = await crypto.subtle.digest(
+        'SHA-256',
+        new TextEncoder().encode(assetIdHex),
+      );
+      const hue = getHueFromHash(hashBuffer);
+
+      const percentage = (value / totalValue) * 100;
+
+      return {
+        percentage,
+        color: `hsl(${hue}, 70%, 50%)`,
+        balance,
+        hasError,
+      };
+    }),
+  );
+
+  // Sort by value percentage in descending order
+  const sorted = distributionWithMetadata.sort((a, b) => b.percentage - a.percentage);
+
+  // Group small assets (less than 2%) into "Other" category for distribution only
+  const SMALL_ASSET_THRESHOLD = 2;
+  const mainAssets = sorted.filter(asset => asset.percentage >= SMALL_ASSET_THRESHOLD);
+  const smallAssets = sorted.filter(asset => asset.percentage < SMALL_ASSET_THRESHOLD);
+
+  const otherPercentage = smallAssets.reduce((acc, asset) => acc + asset.percentage, 0);
+
+  const distributionWithOther = [
+    ...mainAssets.map(({ percentage, color, hasError }) => ({
+      percentage,
+      color,
+      hasError,
+      isOther: false as const,
+    })),
+    // Only add Other if there are small assets
+    ...(otherPercentage > 0
+      ? [
+          {
+            percentage: otherPercentage,
+            color: 'hsl(0, 0%, 50%)', // Gray color for Other
+            hasError: false,
+            isOther: true as const,
+          },
+        ]
+      : []),
+  ];
+
+  return {
+    distribution: distributionWithOther,
+    sortedBalances: sorted.map(({ balance }) => balance), // Keep all assets in the table
+  };
+};
+
+export const AssetsTable = observer(() => {
+  const router = useRouter();
+  const { connected } = connectionStore;
+  const addressIndex = new AddressIndex({ account: connectionStore.subaccount });
+  const { data: balances, isLoading: balancesLoading } = useBalances(addressIndex);
+  const { data: assets, isLoading: assetsLoading } = useAssets();
+  const { data: chainId } = useChainId();
+  const [distribution, setDistribution] = useState<{
+    distribution: {
+      percentage: number;
+      color: string;
+      hasError: boolean;
+      isOther: boolean;
+    }[];
+    sortedBalances: BalancesResponse[];
+  }>();
+
+  useEffect(() => {
+    if (balances) {
+      void calculateAssetDistribution(balances).then(setDistribution);
+    }
+  }, [balances]);
+
+  const isTestnet = chainId !== 'penumbra-1';
+  const stableCoinSymbol = isTestnet ? 'UM' : 'USDC';
+
+  if (!connected) {
+    return <NotConnectedNotice />;
+  }
```

Comment:
suggestion: display loading screen before 'not connected' notice, so the page doesn't flicker too much. This means that loading variable should be true on the server, where we can't check the connection state